### ISSUE TITLE
`misc-const-correctness` false positive with non-trivial object which is being moved

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#57266]
```cpp
#include <list>
#include <string>

class A {
    std::string s;

public:
    A();
    ~A(); // remove to fix warning
};

static void f()
{
    std::list<A> l;

    A a;
    l.emplace_back(std::move(a));
}
```

```
<source>:16:5: warning: variable 'a' of type 'A' can be declared 'const' [misc-const-correctness]
    A a;
    ^
      const
```

https://godbolt.org/z/sf5ox1arK

**Comment by rymiel** [comment]
In the pedantic sense, clang-tidy is correct. Since you've declared a destructor, the implicit move constructor isn't declared as per language rules. This can be [observed on cppinsights.io](https://cppinsights.io/s/b08db70d).
As such, the emplace actually invokes the implicit copy constructor, which takes a const argument.
This can also be observed by explicitly deleting the copy constructor, which makes the code [fail during semantic checking](https://godbolt.org/z/eedzcc4Ts)

**Comment by firewave** [comment]
That sounds vaguely familiar.  But I might be thinking of `-Wdeprecated-copy-with-user-provided-dtor` and `-Wdeprecated-copy-with-dtor`.

FYI the fixed code triggers `performance-move-const-arg` for the `std::move()`. That's how I came across this.

**Comment by JonasToth** [comment]
> That sounds vaguely familiar. But I might be thinking of `-Wdeprecated-copy-with-user-provided-dtor` and `-Wdeprecated-copy-with-dtor`.
> 
> FYI the fixed code triggers `performance-move-const-arg` for the `std::move()`. That's how I came across this.
> 
> ```
> std::move of the const variable 'decl' has no effect or make the variable non-const [performance-move-const-arg,-warnings-as-errors]
> ```
> 
> Would be possible/make sense to have a check which warns about that non-move? Because if you apply the fixes for those checks it is unlikely that you will detect the (possibly) unnecessary copy later on.

that sounds like `performance-move-const-arg` does not properly model if the code "actually moves". Not sure which of those two checks should get a fix. I tend more towards `performance`, on the other hand, this is really non-obvious and confusing.

maybe the `ExprMutAnalyzer` should stay as is, but the clang-tidy portion could match on separate `std::move` calls and consider them as mutation?! not so nice either ...

**Comment by vbvictor** [comment]
As I understand from discussion: const-correctness check is technically correct and there is nothing to fix in that check. Tracking of "misleading moves" should be done separately by other check.

Feel free to reopen issue and continue discussion if I missed something
